### PR TITLE
first complete mt and et version

### PR DESCRIPTION
**Pull request by jbrands**

### CODE REVIEW

File: interface/WrapperTH3D.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+{
+
+    public:
+        WrapperTH3D():IFunctionWrapper() {};
+        WrapperTH3D(const TH3D& h, const std::string& name):IFunctionWrapper(name),m_histo(h) {};
+        virtual ~WrapperTH3D();
+
+        double value(size_t size, const double* xs) override
+        {
+            if(size<2) return 0.;
```
Comment (jbsauvan) [inline interface/WrapperTH3D.h]: This should be size<3

File: test/producePublicFakeFactors.py [hunk 2]
```diff
@@ -283,496 +387,59 @@
...
-fill(ff_qcd_ss, qcd_ss_up,   sys='ff_qcd_up')
-fill(ff_qcd_ss, qcd_ss_down, sys='ff_qcd_down')
 fill(ff_qcd_os, qcd_os)
-fill(ff_qcd_os, qcd_os_up,   sys='ff_qcd_up')
-fill(ff_qcd_os, qcd_os_down, sys='ff_qcd_down')
+fill(ff_qcd_os, qcd_os_up,   sys='ff_qcd_syst_up')
+fill(ff_qcd_os, qcd_os_down, sys='ff_qcd_syst_down')
+fill(ff_qcd_os, qcd_os_up_stat,   sys='ff_qcd_stat_up')
+fill(ff_qcd_os, qcd_os_down_stat, sys='ff_qcd_stat_down')
+fill(ff_qcd_os, frac_qcd, sys='frac_qcd')
```
Comment (jbsauvan) [inline test/producePublicFakeFactors.py]: What is this systematic? I don't understand why you are replacing ff_qcd_os with {frac_qcd}+0*{ff_qcd_os} (so = frac_qcd).

File: test/producePublicFakeFactors.py [hunk 3]
```diff
@@ -283,496 +387,59 @@
...
 fill(ff_w     , w)
-fill(ff_w, w_up,   sys='ff_w_up')
-fill(ff_w, w_down, sys='ff_w_down')
+fill(ff_w, frac_w, sys='frac_w')
+fill(ff_w, frac_dy, sys='frac_dy')
```
Comment (jbsauvan) [inline test/producePublicFakeFactors.py]: ditto

File: test/producePublicFakeFactors.py [hunk 4]
```diff
@@ -283,496 +387,59 @@
...
 fill(ff_tt    , tt)
-fill(ff_tt, tt_corr_up,   sys='ff_tt_corr_up')
-fill(ff_tt, tt_corr_down, sys='ff_tt_corr_down')
-fill(ff_tt, tt_stat_up,   sys='ff_tt_stat_up')
-fill(ff_tt, tt_stat_down, sys='ff_tt_stat_down')
+fill(ff_tt, tt_up,   sys='ff_tt_syst_up')
+fill(ff_tt, tt_down, sys='ff_tt_syst_down')
+fill(ff_tt, tt_up_stat,   sys='ff_tt_stat_up')
+fill(ff_tt, tt_down_stat, sys='ff_tt_stat_down')
+fill(ff_tt, frac_tt, sys='frac_tt')
```
Comment (jbsauvan) [inline test/producePublicFakeFactors.py]: ditto

File: test/producePublicFakeFactors.py [hunk 5]
```diff
@@ -283,496 +387,59 @@
...
 fill(ff_comb  , comb)
-fill(ff_comb, comb_qcd_up,   sys='ff_qcd_up')
```
Comment (jbsauvan) [inline test/producePublicFakeFactors.py:751]: I don't think it will work if you don't register the systematics in the combined fake factor

**Comment by jbsauvan** [comment]
@jbrands @mflechl 
I will perform some tests later Today.
First, a quick question: is there a reason why uncertainties_QCD_W.root files are so large (almost 100MB). Is it impacting the size of the fake factor object, or is this size more reasonable?

**Comment by jbsauvan** [comment]
@jbrands @mflechl 
So I did a quick check. The systematics are not correctly registered in the combined fake factor.
You see that if you do for instance `ff_comb.value( inputs,"frac_qcd" )`

**Comment by jbrands** [comment]
We have systematics for every component of the fake factor. But we don't have a combined systematic so far. This is the reason why analyzers should also be able to extract the 4 different fractions. So in order to retrieve the systematics one has to get e.g. ff_qcd_os->value(inputs,"ff_qcd_syst_up"). We did the combination of all systematics in our code via toys but it can also be done by e.g. (ff_qcd+syst_qcd_up)_frac_qcd+ff_w_(frac_w+frac_dy)+ff_tt*frac_tt -> But you are right I should include all these systematics in the ff_comb

**Comment by jbsauvan** [comment]
- If you have systematics for each of the components it is straightforward to have them in the combined fake factors, by doing the same replacements in `ff_comb` as in your individual components.
- Analyzers should not bother with combining fake factors, etc. They must be able to take the combined fake factor and apply systematics out-of-the-box.

**Comment by jbrands** [comment]
To be honest, I just figured it out that your framework already provides the possibility to get the systematics for ff_comb like for the individual components. I didn't really touch that part of the code so far. But you are absolutely right, this is way more elegant and better for the analyzers! Thx a lot

**Comment by jbrands** [comment]
I updated the pull request, all 3 components (qcd,w,tt) have combined syst uncertainties (from the corrections) and statistical uncertainties on the raw ff. Indeed the QCD_W uncertainties file is quite large. I will try to minimize it. I also updated the twiki.

File: test/producePublicFakeFactors.py [hunk 6]
```diff
@@ -3,170 +3,156 @@
...
     }
 )
 
+###########################################################################################################
 ### W fake factors
+
 w = Node(
     name='ff_w',
-    formula='{mtcorr_w}*{mviscorr_w}*{ff_raw_w}',
+    formula='{mtcorr_w}*{ff_raw_w}+0*{mviscorr_w}',
```
Comment (jbsauvan) [inline test/producePublicFakeFactors.py]: Is the `+0*{mviscorr_w}` what you want? If this is the case it is probably better to just remove the mvis correction.

File: test/producePublicFakeFactors.py [hunk 7]
```diff
@@ -3,170 +3,156 @@
 import os
 
 ## Meta-data
-version = '20160809'
-tag     = 'v0.1.3'
+version = '20160913'
+channel= 'et'
```
Comment (jbsauvan) [inline test/producePublicFakeFactors.py]: Probably better to implement a loop on all final states and categories, in order to produce the fake factors all together.

**Comment by jbsauvan** [comment]
I notice that there is no uncertainty on the fractions included. Are they neglected here because the impact on the combined fake factor is tiny?

**Comment by jbrands** [comment]
You are right, I forgot to comment in the mvis correction vor W fake factor. The impact is not very large but still very good that you noticed that.

**Comment by jbrands** [comment]
We didn't take the uncertainties on the fractions into account at the moment, so I thought it is maybe the best to remove it from the producePublicFakeFactors.py file in order not to confuse analyzers

**Comment by jbrands** [comment]
The uncertainty file combines for QCD all three uncertainties in a 3D histo. Since every uncertainty has 500 points, this gets quite large. For the moment it is the best to just not copy the uncertainty file into every category. So there is just 1 uncertainty file for mt and et which is in the mt and et branch in the same folder as the categories. This reduces the size from 1.2 GB to 100 MB

**Comment by jbrands** [comment]
But this is just temporary, I will try to reduce the size of the uncertainty file for the next commit.
Since I did changed the mvis correction I uploaded new files in /afs/cern.ch/user/j/jbrandst/public/Htautau/ and also changed the version to 20160914 at the twiki

**Comment by jbsauvan** [comment]
Thanks

> So there is just 1 uncertainty file for mt and et which is in the mt and et branch in the same folder as the categories. This reduces the size from 1.2 GB to 100 MB

Actually I was mostly worried about the size of the final fake factor file, but it seems to be much more reasonable (less than 10MB)

> We didn't take the uncertainties on the fractions into account at the moment, so I thought it is maybe the best to remove it from the producePublicFakeFactors.py file in order not to confuse analyzers

Ok, I agree.

I will go again one last time through the code and then merge it. 
Do you have write access to this repository (for pushing the new tag)? If not I or Martin can create the tag (this is `v0.2.1` right?).

File: test/producePublicFakeFactors.py [hunk 8]
```diff
@@ -2,780 +2,591 @@
...
+categories = ['incl','_0jet','_1jet','_1jetZ050','_1jetZ50100','_1jetZ100','_2jet','_2jetVBF','_anyb']
 
-# Individual fake factors
-ff_qcd_os = FakeFactor(vars=['tau_pt', 'tau_decay', 'mvis', 'mu_iso'])
-ff_qcd_ss = FakeFactor(vars=['tau_pt', 'tau_decay', 'mvis', 'mu_iso'])
-ff_w      = FakeFactor(vars=['tau_pt', 'tau_decay', 'mvis', 'mt'])
-ff_tt     = FakeFactor(vars=['tau_pt', 'tau_decay', 'mvis'])
-# Combined fake factor
-ff_comb   = FakeFactor(vars=['tau_pt', 'tau_decay', 'mvis', 'mt', 'mu_iso'])
+for x in range(0,len(categories)):
```
Comment (jbsauvan) [inline test/producePublicFakeFactors.py]: Probably easier to read (and more pythonic):

``` python
for category in categories:
```

File: test/producePublicFakeFactors.py [hunk 9]
```diff
@@ -2,780 +2,591 @@
 import ROOT
 import os
 
-## Meta-data
-version = '20160809'
-tag     = 'v0.1.3'
+#Meta-data
+channel="et"
```
Comment (jbsauvan) [inline test/producePublicFakeFactors.py]: Don't you want also to loop on `et` and `mt` channels?

**Comment by jbrands** [comment]
It would be nice if you can push the tag. It's v0.2.1, you are right!
Thanks a lot for all the comments!

**Comment by jbsauvan** [comment]
@jbrands 
Tag `v0.2.1` has been added:
https://github.com/CMS-HTT/Jet2TauFakes/tree/v0.2.1

**Pull request merged**